docs/design/phases.md

```diff
@@ -0,0 +1,354 @@
+ - Leader election is enabled for both the controller-manager and the scheduler
+ - `HostNetwork: true` is present on all static pods since there is no network configured yet
+ - The yaml manifests are located in `/etc/kubernetes/manifests`
+
+### Addons
+
+#### mark-master
+
+This addon essentially does just this:
+
+```bash
+kubectl taint node ${master_name} node-role.kubernetes.io/master:NoSchedule
+kubectl label node ${master_name} node-role.kubernetes.io/master=""
+```
+
+#### cluster-info
+
+This phase creates the `cluster-info` ConfigMap in the `kube-public` namespace as defined in https://github.com/kubernetes/community/blob/master/contributors/design-proposals/bootstrap-discovery.md
+ - The `ca.crt` and the address/port of the apiserver is added to the `cluster-info` ConfigMap in the `kubeconfig` key
+ - Exposes the `cluster-info` ConfigMap to unauthenticated users (i.e. users in RBAC group `system:unauthenticated`)
+
+#### self-hosting
+
+Parses the yaml static pod manifests in `/etc/kubernetes/manifests` and converts them to native Deployments and DaemonSets:
+ - TODO
+
+#### kube-proxy
+
+A ServiceAccount for `kube-proxy` is created in the `kube-system` namespace.
+
+Deploy kube-proxy as a DaemonSet:
+ - the credentials (`ca.crt` and `token`) to the master come from the ServiceAccount
+ - the location of the master comes from a ConfigMap
+ - the `kube-proxy` ServiceAccount is bound to the privileges in the `system:node-proxier` ClusterRole
+
+#### kube-dns
+
+A ServiceAccount for `kube-dns` is created in the `kube-system` namespace.
+
+Deploy the kube-dns Deployment and Service:
+ - it's the upstream kube-dns deployment relatively unmodified
+ - the `kube-dns` ServiceAccount is bound to the privileges in the `system:kube-dns` ClusterRole
+
+the TLS Bootstrap role (`system:node-bootstrapper`), so tokens in the group `kubeadm:kubelet-bootstrap` only can access what's required for TLS Bootstrapping
+
+## `kubeadm phase` UX
+
+ - `kubeadm phase certs selfsign`
+   - Inputs:
+     - `--dns-domain=cluster.local`
+     - `--service-cidr=10.96.0.1/12`
+     - `--alt-names=""`
+     - `--cert-dir=/etc/kubernetes/pki`
+   - Outputs (in the `--cert-dir` directory):
+     - `ca.{crt,key}`
+     - `apiserver.{crt,key}`
+     - `apiserver-kubelet-client.{crt,key}`
+     - `sa.{pub,key}`
+     - `front-proxy-ca.{crt,key}`
+     - `front-proxy-client.{crt,key}`
+ - `kubeadm phase kubeconfig clientcert`
+   - Inputs:
+     - `--client-name=`
+     - `--api-servers=`
+     - `--cert-dir=/etc/kubernetes/pki`
+     - `--organization=""`
+   - Output:
+     - The kubeconfig is `stdout`
+ - `kubeadm phase kubeconfig token`
+   - Inputs:
+     - `--client-name=`
+     - `--api-servers=`
+     - `--token`
+     - `--cert-dir=/etc/kubernetes/pki`
+   - Output:
+     - The kubeconfig is `stdout`
+ - `kubeadm phase controlplane etcd`
+   - Inputs:
+     - `--image=gcr.io/google_containers/etcd-${GOARCH}:3.0.14-kubeadm`
+     - `--data-dir=/var/lib/etcd`
```

Comment:
By extending etcd flags, it is possible to provide with a minimal effort an early support for creating HA clusters, e.g.:
1. With `kubeadmin init`, create the first master with 
   - `--apiserver-advertise-address` and `--cert-altnames` properly set
   - an etcd.yaml that contains instructions for spinning up an etcd instance that should be member of an etcd cluster
2. Manually copy `/etcd/kubernetes` folder to a second and third masters (probably also necessary to replace kubelet.conf with a new one containing new credentials, created with `kubeconfig user`)
3. Manually setup a LB/DNS in front of all api-servers (matching with `--apiserver-advertise-address` and `--cert-altnames`)
4. Join nodes 

In order to support this a possibile solution is:
- add `--name` flag to etcd command, and assign to it a unique name (e.g. the nodename value retrived from Downward API)
- add a `--etcd-cluster-discovery` flag supporting a dynamic approach for setting up an etcd cluster (no need of knowing the name/ip of nodes in advance, but access to internet/to another etcd instance required)
- add a `--etcd-cluster-members' flag supporting a static approach for setting up an etcd cluster (required knowing the name/ip of nodes in advance, no need of access to internet/to another etcd instance)